### PR TITLE
test(ui-color-picker): add test to `ColorMixer`

### PR DESCRIPTION
**Pull request by NamNguyenInstructure**

### CODE REVIEW

File: packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx [hunk 1]
```diff
@@ -128,4 +146,1591 @@ describe('<ColorMixer />', () => {
...
+          />
+        )
+        const component = await ColorMixerLocator.find()
+        const hueSlider = await component.findColorSlider()
+        const rect = hueSlider.getBoundingClientRect()
+        await hueSlider.mouseDown({ clientX: rect.x + rect.width })
+        expect(onChange).to.have.been.called()
+      })
+
+      it('click at the begin of the slider', async () => {
```
Comment (matyasf) [inline packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx:349]: typo "click at the beginning of the slider"

File: packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx [hunk 2]
```diff
@@ -128,4 +146,1591 @@ describe('<ColorMixer />', () => {
...
+            {...testInputLabels}
+            {...testScreenReaderLabels}
+            onChange={onChange}
+          />
+        )
+        const component = await ColorMixerLocator.find()
+        const hueSlider = await component.findColorSlider()
+        const rect = hueSlider.getBoundingClientRect()
+        await hueSlider.mouseDown({ clientX: rect.x })
+        expect(onChange).to.have.been.calledTwice()
```
Comment (matyasf) [inline packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx:363]: Why is this called twice? Could this be a bug?

File: packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx [hunk 3]
```diff
@@ -128,4 +146,1591 @@ describe('<ColorMixer />', () => {
...
+            {...testInputLabels}
+            {...testScreenReaderLabels}
+            onChange={onChange}
+          />
+        )
+        const component = await ColorMixerLocator.find()
+        const hueSlider = await component.findColorSlider()
+        const rect = hueSlider.getBoundingClientRect()
+        await hueSlider.mouseDown({ clientX: rect.x - 100 })
+        expect(onChange).to.have.been.calledTwice()
```
Comment (matyasf) [inline packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx:380]: Why is this called twice? Could this be a bug?

File: packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx [hunk 4]
```diff
@@ -128,4 +146,1591 @@ describe('<ColorMixer />', () => {
...
+            {...testInputLabels}
+            {...testScreenReaderLabels}
+            onChange={onChange}
+          />
+        )
+        const component = await ColorMixerLocator.find()
+        const hueSlider = await component.findColorSlider()
+        const rect = hueSlider.getBoundingClientRect()
+        await hueSlider.mouseDown({ clientX: rect.x + rect.width + 100 })
+        expect(onChange).to.have.been.calledTwice()
```
Comment (matyasf) [inline packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx:397]: Why is this called twice? Could this be a bug (there are other tests with the same weirdness)?

File: packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx [hunk 5]
```diff
@@ -128,4 +146,1591 @@ describe('<ColorMixer />', () => {
...
+      })
+    })
+
+    it("hue value change with 'a' key pressed", async () => {
```
Comment (matyasf) [inline packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx:401]: I'd put these the "hue value change with 'xyz' key pressed" tests to a single test with a for loop

File: packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx [hunk 6]
```diff
@@ -128,4 +146,1591 @@ describe('<ColorMixer />', () => {
...
+      await hueSlider.keyDown('d')
+      const pos_1 = indicator.getBoundingClientRect()
+      await hueSlider.keyDown('d')
+      const pos_2 = indicator.getBoundingClientRect()
+      expect(pos_2.x).to.be.eq(pos_1.x)
+    })
+  })
+
+  describe('alpha slider', () => {
+    it('should not call onChange when receive `tab` key press', async () => {
```
Comment (matyasf) [inline packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx:633]: typo: "should not call onChange when a `tab` key press is received"

File: packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx [hunk 7]
```diff
@@ -128,4 +146,1591 @@ describe('<ColorMixer />', () => {
...
+        />
+      )
+
+      const component = await ColorMixerLocator.find()
+      const alpharSlider = await component.findAlphaSlider()
+      await alpharSlider.keyDown('tab')
+      expect(onChange).to.not.have.been.called()
+    })
+
+    describe('alpha value change when receive a mousedown event', () => {
```
Comment (matyasf) [inline packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx:650]: typo: "alpha value should change when a mousedown event is received"

File: packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx [hunk 8]
```diff
@@ -128,4 +146,1591 @@ describe('<ColorMixer />', () => {
...
+      )
+
+      const component = await ColorMixerLocator.find()
+      const alpharSlider = await component.findAlphaSlider()
+      expect(await alpharSlider.find('[disabled]')).to.exist()
+    })
+  })
+
+  describe('color palette', () => {
+    describe('should change the color when receive mousedown', () => {
```
Comment (matyasf) [inline packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx:989]: typo: "should change the color when mousedown event is received"

File: packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx [hunk 9]
```diff
@@ -128,4 +146,1591 @@ describe('<ColorMixer />', () => {
...
+          />
+        )
+        const component = await ColorMixerLocator.find()
+        const colorPalette = await component.findColorPalette()
+        const rect = await colorPalette.getBoundingClientRect()
+        await colorPalette.mouseDown({
+          clientX: rect.x + rect.width / 2,
+          clientY: rect.y + rect.height / 2
+        })
+        expect(onChange).to.have.been.called()
```
Comment (matyasf) [inline packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx:1006]: I'd also test for the parameters of the `onChange` function. For examples look for tests using `calledWithMatch` or `calledWith`

File: packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx [hunk 10]
```diff
@@ -128,4 +146,1591 @@ describe('<ColorMixer />', () => {
...
+              {...testInputLabels}
+              {...testScreenReaderLabels}
+              onChange={onChange}
+            />
+          )
+
+          const component = await ColorMixerLocator.find()
+          const palette = await component.find('[aria-label*="color palette"]')
+          await palette.keyDown(keyEvent)
+          // use `calledTwice` because it is called first time when passing  `testValue` color
```
Comment (matyasf) [inline packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx:1194]: such comments are very useful for future readers!

File: packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx [hunk 11]
```diff
@@ -128,4 +146,1591 @@ describe('<ColorMixer />', () => {
...
+        '[class*="ColorPalette__indicator"]'
+      )
+      const pos_1 = await indicatorRef.getBoundingClientRect()
+      await palette.keyDown('ArrowRight')
+      const pos_2 = await indicatorRef.getBoundingClientRect()
+      expect(pos_2.x).to.be.gt(pos_1.x)
+      expect(onChange).to.have.been.calledTwice()
+    })
+
+    it('the indicator moves up when receive the "w" keyboard event', async () => {
```
Comment (matyasf) [inline packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx:1295]: You should make one test, that tests for the "up" key and the "w" key.

File: packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx [hunk 12]
```diff
@@ -128,4 +146,1591 @@ describe('<ColorMixer />', () => {
...
+      )
+      const component = await ColorMixerLocator.find()
+      const RGBAInput = await component.find('[class*="RGBAInput"]')
+      const alphaInput = await RGBAInput.find('[class*="-RGBAInput__aInput"]', {
+        expectEmpty: true
+      })
+      expect(alphaInput).to.not.exist()
+    })
+
+    it('should not call oncChange when `disabled` is set and get the input', async () => {
```
Comment (matyasf) [inline packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx:1611]: typo in oncChange

**Review by matyasf — COMMENTED** [review]
Very nice work! Generally I love that its very thorough and that you tested lots of edge cases too. 
While its important to have atomic tests, I think its OK to merge some very similar ones (e.g. "up" and "W" do the same thing").  Also its a good practice to leave comments in the tricky parts of your code for future readers so they understand it better.

File: packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx (hunk 2 above)
Comment (NamNguyenInstructure) [reply packages/ui-color-picker/src/ColorMixer/__tests__/ColorMixer.test.tsx:363]: Because as the default, the indicator was at the beginning of the slider then I had to move the indicator to somewhere in the middle of the slider by passing them a random value, then I click to move it back beginning so it will change the color twice, once when I pass the `props.value` and once when I click.

**Review by HerrTopi — APPROVED** [review]
